# Kenya Financial Inclusion Analysis 2024
## Survey Coverage and Data Quality

This section presents an assessment of the survey coverage and data quality underlying the FinAccess 2024 analytical view, with the aim of establishing the reliability of the dataset prior to analysis. The evaluation focuses on the respondent base, county coverage, demographic coding, and data completeness. Specifically, it confirms that the dataset contains valid and non-duplicated respondent records, that all 47 counties are adequately represented for meaningful comparison, and that key demographic variables such as gender, age, and income are consistently coded to support accurate segmentation. In addition, the extent of missing values across critical financial indicators is examined to determine the overall completeness of the data. These checks provide a necessary foundation for ensuring that subsequent analysis on county, demographic, and financial access patterns is both credible and statistically reliable.



In [1]:
# Shared setup keeps credentials, paths, and database access consistent across the analysis.
from pathlib import Path
import os

import pandas as pd
from sqlalchemy import create_engine, text

# Credential loading keeps local secrets out of versioned analysis files.
# Fallback loading keeps the connection usable across different local kernels.
try:
    from dotenv import load_dotenv
except ImportError:
    def load_dotenv(dotenv_path):
        dotenv_path = Path(dotenv_path)
        if not dotenv_path.exists():
            return False
        for line in dotenv_path.read_text(encoding="utf-8").splitlines():
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, value = line.split("=", 1)
            os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))
        return True

# Absolute paths prevent exports from depending on the active working directory.
PROJECT_ROOT = Path(r"C:\Portfolio\FinancialInclusion2024")
DATA_DIR = PROJECT_ROOT / "data"
OUTPUT_DIR = PROJECT_ROOT / "outputs"
FIGURE_DIR = OUTPUT_DIR / "figures"

OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
FIGURE_DIR.mkdir(parents=True, exist_ok=True)

# The local .env file keeps the database password outside analysis outputs.
# Git ignores .env so credentials remain local to the analyst machine.
load_dotenv(PROJECT_ROOT / ".env")
DB_PASSWORD = os.getenv("DB_PASSWORD")

if not DB_PASSWORD:
    raise ValueError(f"DB_PASSWORD is missing. Add it to {PROJECT_ROOT / '.env'} before running this notebook.")

DATABASE_URL = f"postgresql+psycopg2://postgres:{DB_PASSWORD}@localhost:5432/finaccess_db"
engine = create_engine(DATABASE_URL)

# Database validation prevents downstream analysis from using the wrong connection.
pd.read_sql("SELECT current_database() AS database_name, current_user AS user_name;", engine)


,database_name,user_name
0,finaccess_db,postgres


## Section 1: Survey Coverage and Core Variables

This section presents a profile of the FinAccess 2024 analytical view used in the financial inclusion analysis, with the aim of establishing the adequacy and reliability of the dataset prior to interpretation. The section confirms the respondent base and verifies that the dataset contains valid and non-duplicated observations, while also assessing the distribution of respondents across counties to ensure adequate national representation, with particular attention to Nairobi coverage. In addition, it examines the structure and consistency of demographic coding, including gender and age categories, as well as the overall age distribution of respondents. Further, the section verifies the availability and completeness of core variables related to financial access, usage, and financial health, which are essential for analysis. These checks provide a necessary foundation for ensuring that subsequent county-level, demographic, and product-level findings are accurate, consistent, and reliable.


In [2]:
# Database readiness is checked before respondent and county summaries are produced.
print("Database engine ready:", engine.url.database)

Database engine ready: finaccess_db


### 1. Master View Preview

The master view consolidates key variables required for analysis into a single structured table. It integrates respondent identifiers, county location, demographic codes, financial access indicators, financial health scores, and vulnerability measures. This unified structure eliminates the need for multiple joins during analysis and ensures consistency across variables. A brief preview of the dataset is undertaken to confirm that the records are complete, correctly aligned, and representative of the survey structure. This step verifies that the analytical view is properly constructed and suitable for both national-level summaries and county-level comparisons.


In [3]:
# Initial sample verifies that respondent, geography, access, and outcome fields are present.
query = """
SELECT *
FROM finaccess_master
LIMIT 10;
"""

df_preview = pd.read_sql(query, engine)
df_preview

,interview__key,county,Sex,Age,Education,Marital,Latitude,Longitude,Quintiles,indWeight,...,Savings_usage,Loan_usage,All_Insurance_including_NHIF,Informal_usage,mobile,digital_acc,Overall_Access_fnl,Financial_literacy_index_fnl,finhealth_fnl,vul_index_fnl
0,08-70-53-89,23.0,2.0,4.0,1.0,3.0,4.197839,34.388253,2.0,382.547882,...,1.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0,0.0,2.0
1,57-84-27-82,35.0,1.0,5.0,3.0,4.0,-0.450037,35.194234,4.0,1606.605713,...,1.0,0.0,1.0,1.0,1.0,1.0,1.0,2.0,0.0,3.0
2,89-63-94-01,28.0,2.0,6.0,2.0,3.0,1.291664,35.675510,3.0,855.210938,...,1.0,0.0,0.0,0.0,1.0,1.0,1.0,1.0,0.0,2.0
3,66-05-72-31,12.0,2.0,4.0,3.0,4.0,-0.018332,37.584181,2.0,763.181152,...,0.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0,0.0,2.0
4,58-45-46-32,15.0,1.0,3.0,2.0,1.0,-1.355414,38.008732,4.0,484.132446,...,0.0,1.0,0.0,1.0,0.0,0.0,2.0,2.0,0.0,3.0
5,66-78-40-09,39.0,1.0,6.0,2.0,4.0,0.835253,34.955861,1.0,1042.621338,...,1.0,1.0,0.0,1.0,1.0,1.0,1.0,2.0,0.0,2.0
6,59-09-18-98,47.0,1.0,3.0,4.0,1.0,-1.289697,36.890205,5.0,3223.651611,...,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,3.0
7,83-61-19-50,6.0,1.0,4.0,2.0,4.0,-3.385262,38.567237,4.0,892.315063,...,1.0,1.0,1.0,0.0,1.0,1.0,1.0,2.0,1.0,3.0
8,00-69-75-78,28.0,1.0,3.0,4.0,4.0,0.974589,35.562473,4.0,173.195862,...,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,3.0
9,43-59-74-21,45.0,2.0,3.0,3.0,4.0,-0.896313,34.662770,3.0,1135.662476,...,1.0,0.0,1.0,1.0,1.0,1.0,1.0,1.0,0.0,3.0


The sample confirms that each respondent record combines geography, demographics, mobile and formal finance indicators, insurance, informal finance, financial literacy, financial health, and vulnerability. This structure supports county comparisons and segmentation by gender, education, age, and income.


### 2. Core Respondent Profile

The selected fields define the minimum demographic profile required for analysis, including respondent identifier, county, gender, age group, and education level. These variables provide the necessary social and geographic context, allowing financial access outcomes to be interpreted meaningfully across different population segments and locations.


In [4]:
# Core demographic fields anchor later gender, age, education, and county segmentation.
query = """
SELECT
    interview__key,
    county,
    "Sex",
    "Age",
    "Education"
FROM finaccess_master
LIMIT 20;
"""

df_demographics = pd.read_sql(query, engine)
df_demographics

,interview__key,county,Sex,Age,Education
0,08-70-53-89,23.0,2.0,4.0,1.0
1,57-84-27-82,35.0,1.0,5.0,3.0
2,89-63-94-01,28.0,2.0,6.0,2.0
3,66-05-72-31,12.0,2.0,4.0,3.0
4,58-45-46-32,15.0,1.0,3.0,2.0
5,66-78-40-09,39.0,1.0,6.0,2.0
6,59-09-18-98,47.0,1.0,3.0,4.0
7,83-61-19-50,6.0,1.0,4.0,2.0
8,00-69-75-78,28.0,1.0,3.0,4.0
9,43-59-74-21,45.0,2.0,3.0,3.0


The result shows 20 respondent records with a compact demographic profile. That is common in household survey microdata, but it is analytically important: any public-facing chart or written finding should map these codes to official FinAccess labels before final interpretation.


### 3. Survey Coverage

The respondent count provides the base population for all subsequent calculations, including rates, proportions, and comparisons. It defines the denominator used in measuring financial inclusion, exclusion, and product usage across the dataset. Ensuring a stable and adequate national sample is essential for producing reliable and comparable results across counties and demographic groups.


In [5]:
# The national denominator anchors all subsequent rate and percentage calculations.
query = """
SELECT COUNT(*) AS total_respondents
FROM finaccess_master;
"""

df_total_respondents = pd.read_sql(query, engine)
df_total_respondents

,total_respondents
0,20871


The result shows `20,871` total respondents. This confirms that `finaccess_master` preserves the expected survey coverage and appears to contain one row per interview.


### 4. County Sample Distribution

County counts show how records are distributed across Kenya. This is important because county-level interpretation depends on both the observed rate and the number of records behind that rate.


In [6]:
# County counts reveal the sample base behind geographic comparisons.
query = """
SELECT
    county,
    COUNT(*) AS respondent_count
FROM finaccess_master
GROUP BY county
ORDER BY respondent_count DESC;
"""

df_county_counts = pd.read_sql(query, engine)
df_county_counts

,county,respondent_count
0,12.0,637
1,47.0,622
2,37.0,601
3,32.0,572
4,27.0,551
5,22.0,541
6,44.0,531
7,36.0,530
8,3.0,530
9,42.0,526


The result ranks counties by respondent count. In this dataset, the largest county count is county `12` with `637` respondents, followed by Nairobi county `47` with `622` respondents and county `37` with `601` respondents.

For financial inclusion research, this matters because county-level patterns can be compared more responsibly when the analyst understands the number of observations behind each county. Later analysis should use `"indWeight"` for representative estimates, but this unweighted count is a necessary first check.


### 5. Gender Coding

Gender coding is checked before any gender-disaggregated analysis. Confirming the available values supports consistent comparisons across mobile money, banking, savings, and overall access.


In [7]:
# Gender coding must be confirmed before access rates are disaggregated.
query = """
SELECT DISTINCT "Sex"
FROM finaccess_master
ORDER BY "Sex";
"""

df_sex_values = pd.read_sql(query, engine)
df_sex_values

,Sex
0,1.0
1,2.0


The result shows two distinct values: `1` and `2`. This indicates that `"Sex"` is stored as a coded categorical variable in the imported survey data. The result is analytically useful because it confirms that gender-disaggregated analysis is possible, but it also highlights the need to attach official value labels before communicating results to employers, policymakers, or programme teams.


### 6. Nairobi Respondent Base

Nairobi is isolated because it is Kenya's largest urban financial market and often serves as a benchmark for digital and formal financial access. Its respondent base supports a separate county profile.


In [8]:
# Nairobi is isolated for urban financial access benchmarking.
query = """
SELECT
    county,
    COUNT(*) AS nairobi_respondents
FROM finaccess_master
WHERE county = 47
GROUP BY county;
"""

df_nairobi_count = pd.read_sql(query, engine)
df_nairobi_count

,county,nairobi_respondents
0,47.0,622


The result shows that Nairobi, coded as county `47`, has `622` respondents in the analytical view. This gives enough records for an initial Nairobi profile, while also reminding us that county-level statistics should be interpreted with survey design and weighting in mind.


### 7. Age Profile

Age is stored as grouped categories rather than exact years. This supports analysis of youth, working-age adults, and older respondents in relation to financial access and resilience.


In [9]:
# Age-group distribution supports youth, working-age, and older-person segmentation.
query = """
SELECT
    "Age",
    COUNT(*) AS respondent_count
FROM finaccess_master
GROUP BY "Age"
ORDER BY "Age" ASC;
"""

df_age_distribution = pd.read_sql(query, engine)
df_age_distribution

,Age,respondent_count
0,1.0,1121
1,2.0,3997
2,3.0,5361
3,4.0,3942
4,5.0,2545
5,6.0,3905


The largest group is age code `3` with `5,361` respondents, followed by code `2` with `3,997`, code `4` with `3,942`, code `6` with `3,905`, code `5` with `2,545`, and code `1` with `1,121`. This confirms that `"Age"` is a grouped survey variable in the current analytical view.

For financial inclusion analysis, this is valuable because age-band comparisons can still be conducted, but the official codebook should be used to label each age group before presenting insights.


### 8. Additional Respondent Review

A second ordered sample confirms that the respondent-level structure remains consistent beyond the first preview. Stable record review is useful when checking wide survey views.


In [10]:
# Stable respondent ordering supports repeatable record review.
query = """
SELECT *
FROM finaccess_master
ORDER BY interview__key
LIMIT 10 OFFSET 10;
"""

df_rows_11_to_20 = pd.read_sql(query, engine)
df_rows_11_to_20

,interview__key,county,Sex,Age,Education,Marital,Latitude,Longitude,Quintiles,indWeight,...,Savings_usage,Loan_usage,All_Insurance_including_NHIF,Informal_usage,mobile,digital_acc,Overall_Access_fnl,Financial_literacy_index_fnl,finhealth_fnl,vul_index_fnl
0,00-04-15-25,29.0,2.0,4.0,2.0,2.0,0.197719,35.077143,2.0,931.331238,...,1.0,1.0,0.0,0.0,1.0,1.0,1.0,2.0,0.0,2.0
1,00-04-33-79,24.0,1.0,4.0,2.0,4.0,1.246728,35.179556,1.0,938.341614,...,0.0,0.0,0.0,0.0,0.0,0.0,3.0,2.0,0.0,2.0
2,00-04-70-77,45.0,2.0,3.0,4.0,4.0,-0.672814,34.763893,5.0,482.487305,...,1.0,0.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0,3.0
3,00-04-95-13,22.0,2.0,4.0,4.0,4.0,-1.007086,36.751552,4.0,1196.859375,...,1.0,1.0,1.0,1.0,1.0,1.0,1.0,2.0,1.0,2.0
4,00-05-00-35,30.0,2.0,2.0,3.0,4.0,0.271245,35.726968,1.0,2088.838135,...,0.0,0.0,0.0,0.0,1.0,0.0,3.0,2.0,0.0,2.0
5,00-05-23-01,15.0,2.0,6.0,1.0,3.0,-1.665858,37.936409,2.0,896.181274,...,0.0,0.0,0.0,0.0,1.0,1.0,1.0,1.0,0.0,3.0
6,00-05-37-03,22.0,1.0,6.0,4.0,2.0,-1.096728,37.324677,5.0,1120.588135,...,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,3.0
7,00-05-43-82,22.0,1.0,5.0,2.0,4.0,-1.204291,36.689286,3.0,4962.470215,...,1.0,1.0,0.0,1.0,1.0,1.0,1.0,1.0,1.0,3.0
8,00-06-21-46,4.0,2.0,4.0,1.0,4.0,-1.304928,39.661203,1.0,470.096710,...,0.0,1.0,0.0,1.0,0.0,0.0,2.0,3.0,0.0,2.0
9,00-06-29-78,44.0,1.0,4.0,2.0,4.0,-0.772650,34.524960,3.0,2068.205566,...,0.0,1.0,0.0,1.0,1.0,1.0,1.0,3.0,0.0,2.0


The result displays rows 11 through 20 under a stable ordering by `interview__key`. The respondent structure remains consistent beyond the initial preview, reinforcing confidence in the consolidated analytical view.


## Section 2: National and County Access Patterns

This section summarises the respondent distribution across counties, access tiers, gender, education levels, and income quintiles. The aim is to highlight early patterns in financial inclusion across geographic and socioeconomic groups. The results provide an overview of how access to financial services varies by location and demographic characteristics, allowing for initial identification of disparities and concentration of inclusion across the population.


### 1. County Sample Share

County sample shares place respondent counts in national context. This helps identify where county estimates are supported by larger or smaller portions of the sample.


In [11]:
# County sample shares contextualise the strength of geographic comparisons.
query = """
SELECT
    county,
    COUNT(*) AS respondent_count,
    ROUND(
        COUNT(*) * 100.0 / (SELECT COUNT(*) FROM finaccess_master),
        2
    ) AS sample_share_percent
FROM finaccess_master
GROUP BY county
ORDER BY respondent_count DESC;
"""

df_county_sample_share = pd.read_sql(query, engine)
df_county_sample_share

,county,respondent_count,sample_share_percent
0,12.0,637,3.05
1,47.0,622,2.98
2,37.0,601,2.88
3,32.0,572,2.74
4,27.0,551,2.64
5,22.0,541,2.59
6,44.0,531,2.54
7,36.0,530,2.54
8,3.0,530,2.54
9,42.0,526,2.52


The result returns all 47 counties with respondent counts and percentage shares of the 20,871-person sample. County `12` accounts for `3.05%`, Nairobi accounts for `2.98%`, and county `26` accounts for `0.78%`.

For policy and product teams, the finding confirms broad geographic coverage while showing why sample size must be considered alongside county rates.


### 2. Counties with Larger Sample Bases

Counties with more than 500 respondents provide stronger unweighted bases for initial descriptive comparisons. These counties are useful for early profiling before survey-weighted reporting.


In [12]:
# Larger county bases support more detailed subgroup reporting.
query = """
SELECT
    county,
    COUNT(*) AS respondent_count
FROM finaccess_master
GROUP BY county
HAVING COUNT(*) > 500
ORDER BY respondent_count DESC;
"""

df_counties_over_500 = pd.read_sql(query, engine)
df_counties_over_500

,county,respondent_count
0,12.0,637
1,47.0,622
2,37.0,601
3,32.0,572
4,27.0,551
5,22.0,541
6,44.0,531
7,3.0,530
8,36.0,530
9,42.0,526


The result identifies 15 counties with more than 500 respondents. Larger county bases strengthen descriptive confidence, although survey weights and official sample design remain necessary for representative county estimates.


### 3. National Financial Access Tiers

The access-tier classification groups respondents based on their use of financial services, including banking, mobile money, informal finance, or no financial services at all. This categorisation provides a clearer view of financial inclusion by going beyond single indicators and showing the depth and quality of access across the population.


In [13]:
# Access tiers translate product use into policy-relevant inclusion segments.
query = """
WITH access_tiers AS (
    SELECT
        CASE
            WHEN bank_use = 1 AND mobile_money_use = 1 THEN 'Fully Included (bank and mobile money)'
            WHEN mobile_money_use = 1 AND (bank_use IS NULL OR bank_use <> 1) THEN 'Partially Included (mobile only)'
            WHEN "Informal_usage" = 1
                 AND (mobile_money_use IS NULL OR mobile_money_use <> 1)
                 AND (bank_use IS NULL OR bank_use <> 1) THEN 'Informally Served (informal only)'
            ELSE 'Financially Excluded (none)'
        END AS financial_access_tier
    FROM finaccess_master
)
SELECT
    financial_access_tier,
    COUNT(*) AS respondent_count
FROM access_tiers
GROUP BY financial_access_tier
ORDER BY respondent_count DESC;
"""

df_access_tiers = pd.read_sql(query, engine)
df_access_tiers

,financial_access_tier,respondent_count
0,Fully Included (bank and mobile money),9196
1,Partially Included (mobile only),7223
2,Financially Excluded (none),3144
3,Informally Served (informal only),1308


The result shows `9,196` respondents in the fully included tier, `7,223` in the partially included mobile-money-led tier, `3,144` in the financially excluded tier, and `1,308` in the informally served tier. The largest two groups are linked to formal or digital channels, which reflects the central role of mobile money and banking in Kenya's financial inclusion landscape.

For fintechs, the partially included group may represent people who already use mobile money but may not be fully connected to formal banking products, credit, insurance, or savings solutions. For NGOs and development organisations, the excluded and informally served groups point to populations that may need trust-building, consumer protection, lower-cost products, agent access, financial education, or livelihood-linked financial services. Before final publication, this simplified classification should be checked against the official FinAccess codebook and compared with `"Access_Strand"` so that policy-relevant segments align with official survey definitions.


### 4. Gender and Product Usage

Gender-disaggregated rates show whether mobile money, banking, and savings usage differ between male-coded and female-coded respondents. These comparisons are important for inclusive product design and targeted interventions.


In [14]:
# Gender-specific rates reveal whether inclusion differs by respondent group.
query = """
SELECT
    "Sex",
    COUNT(*) AS respondent_count,
    ROUND(AVG(CASE WHEN mobile_money_use = 1 THEN 1.0 WHEN mobile_money_use IS NULL THEN NULL ELSE 0.0 END) * 100, 2) AS mobile_money_use_rate_percent,
    ROUND(AVG(CASE WHEN bank_use = 1 THEN 1.0 WHEN bank_use IS NULL THEN NULL ELSE 0.0 END) * 100, 2) AS bank_use_rate_percent,
    ROUND(AVG(CASE WHEN "Savings_usage" = 1 THEN 1.0 WHEN "Savings_usage" IS NULL THEN NULL ELSE 0.0 END) * 100, 2) AS savings_usage_rate_percent
FROM finaccess_master
GROUP BY "Sex"
ORDER BY "Sex";
"""

df_gender_usage_rates = pd.read_sql(query, engine)
df_gender_usage_rates

,Sex,respondent_count,mobile_money_use_rate_percent,bank_use_rate_percent,savings_usage_rate_percent
0,1.0,8545,78.77,51.79,64.01
1,2.0,12326,78.64,41.18,63.06


The result shows two sex codes. Sex code `1` has `8,545` respondents, while sex code `2` has `12,326`. Mobile money use is very similar across the two groups: `78.77` percent for code `1` and `78.64` percent for code `2`. Savings usage is also close, at `64.01` percent for code `1` and `63.06` percent for code `2`.

The clearest difference is in bank use. Sex code `1` has a bank use rate of `51.79` percent, compared with `41.18` percent for sex code `2`, a gap of about 10.61 percentage points. Once the official value labels are attached, this result can support a gender-focused interpretation of formal banking access. For fintechs, this may point to an opportunity for mobile-first banking products. For NGOs and development organisations, it may signal the need for interventions around documentation, income regularity, account affordability, trust, agent availability, or social norms affecting formal financial access.


### 5. Education and Overall Access

Education categories help assess whether formal schooling is associated with stronger overall financial access. This matters for financial capability programmes and for providers designing low-friction services.


In [15]:
# Education segmentation tests whether access varies with schooling level.
query = """
SELECT
    "Education",
    COUNT(*) AS respondent_count,
    ROUND(AVG("Overall_Access_fnl")::numeric, 2) AS avg_overall_access_fnl
FROM finaccess_master
WHERE "Education" IS NOT NULL
GROUP BY "Education"
ORDER BY "Education";
"""

df_access_by_education = pd.read_sql(query, engine)
df_access_by_education

,Education,respondent_count,avg_overall_access_fnl
0,1.0,3078,1.42
1,2.0,7684,1.30
2,3.0,7345,1.39
3,4.0,2755,1.02
4,5.0,9,1.56


The result shows variation in overall access across education categories, with very small categories requiring caution. Education should remain part of segmentation because financial capability, product comprehension, and trust can differ materially by schooling level.


### 6. Income, Financial Health, and Vulnerability

Income quintiles connect access outcomes to household welfare. This is central to understanding whether inclusion reaches lower-income Kenyans or remains concentrated among better-off households.


In [16]:
# Income gradients link financial access to household welfare and resilience.
query = """
SELECT
    "Quintiles",
    COUNT(*) AS respondent_count,
    ROUND(AVG(finhealth_fnl)::numeric, 2) AS avg_finhealth_fnl,
    ROUND(AVG(vul_index_fnl)::numeric, 2) AS avg_vul_index_fnl
FROM finaccess_master
WHERE "Quintiles" IS NOT NULL
GROUP BY "Quintiles"
ORDER BY "Quintiles" ASC;
"""

df_finhealth_by_quintile = pd.read_sql(query, engine)
df_finhealth_by_quintile

,Quintiles,respondent_count,avg_finhealth_fnl,avg_vul_index_fnl
0,1.0,4973,0.03,2.21
1,2.0,4478,0.07,2.37
2,3.0,4293,0.12,2.48
3,4.0,4232,0.22,2.59
4,5.0,2859,0.39,2.75


The result shows a clear ordered pattern across quintiles. Average `finhealth_fnl` rises from `0.03` in quintile `1` to `0.39` in quintile `5`. The vulnerability index also rises from `2.21` in quintile `1` to `2.75` in quintile `5`.

The financial health pattern suggests that better-off respondents have stronger measured financial health in this coded indicator. The vulnerability index also varies systematically by quintile, although the official codebook should be used to confirm whether a higher `vul_index_fnl` value means greater vulnerability or lower vulnerability. For policy and programme teams, the main implication is that welfare position is strongly related to financial wellbeing. Fintechs designing credit, savings, or insurance products should avoid assuming that access alone produces resilience, while NGOs and development organisations can use quintile-based analysis to prioritise financial capability, safety nets, and consumer protection for lower-welfare groups.


### 7. Completeness of Key Indicators

Completeness checks identify whether the core indicators are reliable enough for national and county analysis. Missingness is especially important for binary access variables because nulls can distort rates.


In [17]:
# Missingness checks protect rate estimates from hidden denominator problems.
query = """
SELECT
    SUM(CASE WHEN interview__key IS NULL THEN 1 ELSE 0 END) AS missing_interview_key,
    SUM(CASE WHEN county IS NULL THEN 1 ELSE 0 END) AS missing_county,
    SUM(CASE WHEN "Sex" IS NULL THEN 1 ELSE 0 END) AS missing_sex,
    SUM(CASE WHEN "Age" IS NULL THEN 1 ELSE 0 END) AS missing_age,
    SUM(CASE WHEN "Education" IS NULL THEN 1 ELSE 0 END) AS missing_education,
    SUM(CASE WHEN "Quintiles" IS NULL THEN 1 ELSE 0 END) AS missing_quintiles,
    SUM(CASE WHEN "indWeight" IS NULL THEN 1 ELSE 0 END) AS missing_indweight,
    SUM(CASE WHEN mobile_money_use IS NULL THEN 1 ELSE 0 END) AS missing_mobile_money_use,
    SUM(CASE WHEN bank_use IS NULL THEN 1 ELSE 0 END) AS missing_bank_use,
    SUM(CASE WHEN mobile_money_access IS NULL THEN 1 ELSE 0 END) AS missing_mobile_money_access,
    SUM(CASE WHEN "Access_Strand" IS NULL THEN 1 ELSE 0 END) AS missing_access_strand,
    SUM(CASE WHEN "Savings_usage" IS NULL THEN 1 ELSE 0 END) AS missing_savings_usage,
    SUM(CASE WHEN "Loan_usage" IS NULL THEN 1 ELSE 0 END) AS missing_loan_usage,
    SUM(CASE WHEN "All_Insurance_including_NHIF" IS NULL THEN 1 ELSE 0 END) AS missing_insurance_including_nhif,
    SUM(CASE WHEN "Informal_usage" IS NULL THEN 1 ELSE 0 END) AS missing_informal_usage,
    SUM(CASE WHEN mobile IS NULL THEN 1 ELSE 0 END) AS missing_mobile,
    SUM(CASE WHEN digital_acc IS NULL THEN 1 ELSE 0 END) AS missing_digital_acc,
    SUM(CASE WHEN "Overall_Access_fnl" IS NULL THEN 1 ELSE 0 END) AS missing_overall_access_fnl,
    SUM(CASE WHEN "Financial_literacy_index_fnl" IS NULL THEN 1 ELSE 0 END) AS missing_financial_literacy_index_fnl,
    SUM(CASE WHEN finhealth_fnl IS NULL THEN 1 ELSE 0 END) AS missing_finhealth_fnl,
    SUM(CASE WHEN vul_index_fnl IS NULL THEN 1 ELSE 0 END) AS missing_vul_index_fnl
FROM finaccess_master;
"""

df_missing_values = pd.read_sql(query, engine)
df_missing_values

,missing_interview_key,missing_county,missing_sex,missing_age,missing_education,missing_quintiles,missing_indweight,missing_mobile_money_use,missing_bank_use,missing_mobile_money_access,...,missing_savings_usage,missing_loan_usage,missing_insurance_including_nhif,missing_informal_usage,missing_mobile,missing_digital_acc,missing_overall_access_fnl,missing_financial_literacy_index_fnl,missing_finhealth_fnl,missing_vul_index_fnl
0,0,0,0,0,0,36,0,6,7,0,...,0,0,0,0,0,0,0,0,0,0


The result shows strong completeness across the main analytical fields. `interview__key`, `county`, `"Sex"`, `"Age"`, `"Education"`, `"indWeight"`, `mobile_money_access`, `"Access_Strand"`, `"Savings_usage"`, `"Loan_usage"`, `"All_Insurance_including_NHIF"`, `"Informal_usage"`, `mobile`, `digital_acc`, `"Overall_Access_fnl"`, `"Financial_literacy_index_fnl"`, `finhealth_fnl`, and `vul_index_fnl` have zero missing values. The variables with missing values are limited: `"Quintiles"` has `36` missing records, `mobile_money_use` has `6`, and `bank_use` has `7`.

This is encouraging for financial inclusion analysis because the key variables are mostly complete. The small missing counts should still be documented, especially when producing rates or quintile comparisons. For fintechs and development organisations, this type of completeness check increases confidence that observed differences are less likely to be driven by missing data.


### 8. Age and Financial Literacy Ranges

The age and literacy ranges confirm how key respondent characteristics are coded. This supports consistent interpretation before comparing groups or building visual summaries.


In [18]:
# Range checks validate coded age and literacy values before interpretation.
query = """
SELECT
    MIN("Age") AS min_age_code,
    MAX("Age") AS max_age_code,
    ROUND(AVG("Age")::numeric, 2) AS avg_age_code,
    MIN("Financial_literacy_index_fnl") AS min_financial_literacy_index_fnl,
    MAX("Financial_literacy_index_fnl") AS max_financial_literacy_index_fnl,
    ROUND(AVG("Financial_literacy_index_fnl")::numeric, 2) AS avg_financial_literacy_index_fnl
FROM finaccess_master;
"""

df_age_literacy_summary = pd.read_sql(query, engine)
df_age_literacy_summary

,min_age_code,max_age_code,avg_age_code,min_financial_literacy_index_fnl,max_financial_literacy_index_fnl,avg_financial_literacy_index_fnl
0,1.0,6.0,3.7,1.0,4.0,1.85


The result shows that `"Age"` ranges from code `1` to code `6`, with an average age code of `3.70`. The financial literacy index ranges from `1` to `4`, with an average value of `1.85`.

For financial inclusion policy, these summaries provide two useful signals. Second, the financial literacy distribution suggests room for deeper analysis of how knowledge and confidence relate to mobile money use, savings, borrowing, insurance, and financial health. For fintechs, this can inform product design and customer education. For NGOs and development organisations, it can support financial capability programming and consumer protection strategies.


## Section 3: Data Quality and Clean Analytical Dataset

This section assesses missingness, confirms core demographic completeness, and creates a clean analytical layer. The goal is to make the county and access indicators consistent for reporting.


### 1. Missing Data Profile

The missing data profile identifies which variables require caution in reporting. Low missingness strengthens confidence in county, gender, income, and access-rate summaries.


In [19]:
# Full-column missingness identifies variables requiring caution in reporting.
query = """
WITH missing_summary AS (
    SELECT 'interview__key' AS column_name, SUM(CASE WHEN interview__key IS NULL THEN 1 ELSE 0 END) AS missing_count, COUNT(*) AS total_count FROM finaccess_master
    UNION ALL SELECT 'county', SUM(CASE WHEN county IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Sex', SUM(CASE WHEN "Sex" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Age', SUM(CASE WHEN "Age" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Education', SUM(CASE WHEN "Education" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Marital', SUM(CASE WHEN "Marital" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Latitude', SUM(CASE WHEN "Latitude" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Longitude', SUM(CASE WHEN "Longitude" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Quintiles', SUM(CASE WHEN "Quintiles" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'indWeight', SUM(CASE WHEN "indWeight" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'mobile_money_use', SUM(CASE WHEN mobile_money_use IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'bank_use', SUM(CASE WHEN bank_use IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'mobile_money_access', SUM(CASE WHEN mobile_money_access IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Access_Strand', SUM(CASE WHEN "Access_Strand" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Savings_usage', SUM(CASE WHEN "Savings_usage" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Loan_usage', SUM(CASE WHEN "Loan_usage" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'All_Insurance_including_NHIF', SUM(CASE WHEN "All_Insurance_including_NHIF" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Informal_usage', SUM(CASE WHEN "Informal_usage" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'mobile', SUM(CASE WHEN mobile IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'digital_acc', SUM(CASE WHEN digital_acc IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Overall_Access_fnl', SUM(CASE WHEN "Overall_Access_fnl" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'Financial_literacy_index_fnl', SUM(CASE WHEN "Financial_literacy_index_fnl" IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'finhealth_fnl', SUM(CASE WHEN finhealth_fnl IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
    UNION ALL SELECT 'vul_index_fnl', SUM(CASE WHEN vul_index_fnl IS NULL THEN 1 ELSE 0 END), COUNT(*) FROM finaccess_master
)
SELECT
    column_name,
    missing_count,
    total_count,
    ROUND(missing_count * 100.0 / total_count, 2) AS missing_percent
FROM missing_summary
ORDER BY missing_percent DESC, missing_count DESC, column_name;
"""

df_missing_summary = pd.read_sql(query, engine)
df_missing_summary

,column_name,missing_count,total_count,missing_percent
0,Quintiles,36,20871,0.17
1,bank_use,7,20871,0.03
2,mobile_money_use,6,20871,0.03
3,Access_Strand,0,20871,0.00
4,Age,0,20871,0.00
5,All_Insurance_including_NHIF,0,20871,0.00
6,county,0,20871,0.00
7,digital_acc,0,20871,0.00
8,Education,0,20871,0.00
9,Financial_literacy_index_fnl,0,20871,0.00


The missing data summary shows that the analytical view is highly complete. `"Quintiles"` has the largest missing count, with `36` missing records, or `0.17` percent of the sample. `bank_use` has `7` missing records and `mobile_money_use` has `6` missing records, each about `0.03` percent of the sample. All other columns in the 24-column view have zero missing values.

This is a strong data quality result. It means the core financial inclusion indicators, demographic fields, and outcome measures can be used with relatively little concern about missingness. The small number of missing values should still be documented because aggregate rates can be affected by how nulls are handled.


### 2. Core Demographic Completeness

County, gender, and age are required for most segmentation in this analysis. Confirming their completeness protects the validity of county and demographic comparisons.


In [20]:
# Core demographic completeness protects county and gender comparisons.
query = """
SELECT COUNT(*) AS clean_core_records
FROM finaccess_master
WHERE county IS NOT NULL
  AND "Sex" IS NOT NULL
  AND "Age" IS NOT NULL;
"""

df_clean_core_records = pd.read_sql(query, engine)
df_clean_core_records

,clean_core_records
0,20871


The result shows `20,871` clean core records, which is the full size of the analytical view. This means no respondents are lost when requiring county, sex, and age to be present.

This is important because it allows later county, gender, and age analysis to use the full respondent base. For fintechs and development organisations, it means that geographic and demographic segmentation can proceed without a major reduction in sample size.


### 3. Working-Age Respondents

Working-age respondents are central to employment, entrepreneurship, household formation, and digital finance adoption. This segment is especially relevant for fintech product design and livelihood programmes.


In [21]:
# Working-age segmentation supports employment, enterprise, and digital finance analysis.
query = """
SELECT
    "Age" AS age_code,
    CASE
        WHEN "Age" = 2 THEN '16-25'
        WHEN "Age" = 3 THEN '26-35'
        WHEN "Age" = 4 THEN '36-45'
        WHEN "Age" = 5 THEN '46-55'
    END AS age_group,
    COUNT(*) AS respondent_count
FROM finaccess_master
WHERE "Age" BETWEEN 2 AND 5
GROUP BY "Age"
ORDER BY "Age";
"""

df_working_age = pd.read_sql(query, engine)
df_working_age

,age_code,age_group,respondent_count
0,2.0,16-25,3997
1,3.0,26-35,5361
2,4.0,36-45,3942
3,5.0,46-55,2545


The working-age result contains four age groups: `16-25` with `3,997` respondents, `26-35` with `5,361`, `36-45` with `3,942`, and `46-55` with `2,545`. The largest group is `26-35`, a segment that is often highly relevant for employment, enterprise activity, household formation, and digital finance adoption.

This filter is analytically justified because financial behaviour differs by life stage. Younger working-age adults may rely heavily on mobile money and digital credit, while older working-age adults may have different savings, insurance, and household responsibility patterns. For fintechs, this supports product segmentation. For NGOs and development organisations, it supports age-sensitive programme design.


### 4. Major Urban Centres

Nairobi, Mombasa, Kisumu, Nakuru, and Uasin Gishu represent major urban markets with different levels of digital and formal financial access. Comparing them helps separate broad urbanisation effects from local market performance.


In [22]:
# Major urban centres benchmark formal and digital access in Kenya's largest markets.
query = """
SELECT
    CAST(county AS INTEGER) AS county_code,
    CASE
        WHEN county = 47 THEN 'Nairobi'
        WHEN county = 1 THEN 'Mombasa'
        WHEN county = 42 THEN 'Kisumu'
        WHEN county = 32 THEN 'Nakuru'
        WHEN county = 27 THEN 'Eldoret / Uasin Gishu'
    END AS urban_centre,
    COUNT(*) AS respondent_count,
    ROUND(AVG(CASE WHEN mobile_money_use = 1 THEN 1.0 WHEN mobile_money_use IS NULL THEN NULL ELSE 0.0 END) * 100, 2) AS mobile_money_use_rate_percent,
    ROUND(AVG(CASE WHEN bank_use = 1 THEN 1.0 WHEN bank_use IS NULL THEN NULL ELSE 0.0 END) * 100, 2) AS bank_use_rate_percent,
    ROUND(AVG(CASE WHEN bank_use = 1 AND mobile_money_use = 1 THEN 1.0 ELSE 0.0 END) * 100, 2) AS fully_included_rate_percent
FROM finaccess_master
WHERE county IN (47, 1, 42, 32, 27)
GROUP BY county
ORDER BY fully_included_rate_percent DESC;
"""

df_urban_inclusion_rates = pd.read_sql(query, engine)
df_urban_inclusion_rates

,county_code,urban_centre,respondent_count,mobile_money_use_rate_percent,bank_use_rate_percent,fully_included_rate_percent
0,47,Nairobi,622,92.93,72.99,72.67
1,27,Eldoret / Uasin Gishu,551,83.48,61.16,59.89
2,32,Nakuru,572,81.12,53.67,52.27
3,1,Mombasa,428,80.80,52.69,51.64
4,42,Kisumu,526,81.94,48.67,47.15


The result shows Nairobi with the highest rates among the selected urban centres: `92.93` percent mobile money use, `72.99` percent bank use, and `72.67` percent fully included by the simplified bank-and-mobile definition. Eldoret through Uasin Gishu follows with a fully included rate of `59.89` percent, then Nakuru at `52.27` percent, Mombasa at `51.64` percent, and Kisumu at `47.15` percent.

The policy implication is that major urban centres are not identical financial markets. Nairobi appears particularly strong on both mobile money and bank use, while Kisumu shows a lower fully included rate among the selected centres despite having high mobile money use. For fintechs, this suggests that go-to-market strategies should not treat all cities as interchangeable. For NGOs and development organisations, it highlights that even urban areas may require differentiated financial capability, trust-building, savings, credit, or insurance interventions.


### 5. Missing Binary Access Indicators

Small numbers of missing mobile money and bank-use values are standardised for cleaned helper fields. This makes later summaries more consistent while preserving the original variables.


In [23]:
# Null binary indicators are standardised to avoid ambiguous access classifications.
query = """
SELECT
    COUNT(*) AS total_records,
    SUM(CASE WHEN mobile_money_use IS NULL THEN 1 ELSE 0 END) AS original_null_mobile_money_use,
    SUM(CASE WHEN bank_use IS NULL THEN 1 ELSE 0 END) AS original_null_bank_use,
    SUM(CASE WHEN COALESCE(mobile_money_use, 0) = 0 THEN 1 ELSE 0 END) AS mobile_money_use_after_coalesce_zero,
    SUM(CASE WHEN COALESCE(bank_use, 0) = 0 THEN 1 ELSE 0 END) AS bank_use_after_coalesce_zero
FROM finaccess_master;
"""

df_coalesce_check = pd.read_sql(query, engine)
df_coalesce_check

,total_records,original_null_mobile_money_use,original_null_bank_use,mobile_money_use_after_coalesce_zero,bank_use_after_coalesce_zero
0,20871,6,7,6,7


The result confirms that `mobile_money_use` has `6` null values and `bank_use` has `7` null values.

This decision should be documented carefully. If the official survey codebook distinguishes missing, refused, not applicable, and no-use responses, then a production analysis may keep those categories separate. For this analysis cleaning view, the use of `COALESCE` creates practical helper columns that prevent null values from causing unexpected behaviour in grouping, dashboards, or simplified classification logic.


### 6. County Code Standardisation

County codes are standardised as integers for cleaner reporting and reliable joins to county reference data. This improves readability in tables and charts.


In [26]:
# Integer county codes improve joins, display, and dashboard compatibility.
query = """
SELECT
    county AS original_county_value,
    CAST(county AS INTEGER) AS county_code_integer,
    COUNT(*) AS respondent_count
FROM finaccess_master
GROUP BY county
ORDER BY county_code_integer
LIMIT 10;
"""

df_county_cast = pd.read_sql(query, engine)
df_county_cast

,original_county_value,county_code_integer,respondent_count
0,1.0,1,428
1,2.0,2,377
2,3.0,3,530
3,4.0,4,333
4,5.0,5,258
5,6.0,6,394
6,7.0,7,306
7,8.0,8,418
8,9.0,9,390
9,10.0,10,369


The result shows that county values can be cleanly displayed as integer county codes. For example, the first ten county codes appear as `1` through `10` after casting, with respondent counts attached to each code.

This is a small but important cleaning step. It prepares the data for clearer county tables and makes it easier to join with external county metadata, shapefiles, or dashboard filters. For organisations working across Kenya's counties, integer county codes and county names are easier to communicate than floating-point identifiers.


### 7. Clean Analysis View

The cleaned view preserves the original survey indicators while adding county names, standardised county codes, and financial access tiers. It provides a consistent reporting layer for later analysis.


In [27]:
# Clean reporting layer centralises county names and access-tier definitions.
from sqlalchemy import text

query = """
CREATE OR REPLACE VIEW finaccess_clean AS
SELECT
    interview__key,
    CAST(county AS INTEGER) AS county,
    CASE
        WHEN CAST(county AS INTEGER) = 1 THEN 'Mombasa'
        WHEN CAST(county AS INTEGER) = 2 THEN 'Kwale'
        WHEN CAST(county AS INTEGER) = 3 THEN 'Kilifi'
        WHEN CAST(county AS INTEGER) = 4 THEN 'Tana River'
        WHEN CAST(county AS INTEGER) = 5 THEN 'Lamu'
        WHEN CAST(county AS INTEGER) = 6 THEN 'Taita Taveta'
        WHEN CAST(county AS INTEGER) = 7 THEN 'Garissa'
        WHEN CAST(county AS INTEGER) = 8 THEN 'Wajir'
        WHEN CAST(county AS INTEGER) = 9 THEN 'Mandera'
        WHEN CAST(county AS INTEGER) = 10 THEN 'Marsabit'
        WHEN CAST(county AS INTEGER) = 11 THEN 'Isiolo'
        WHEN CAST(county AS INTEGER) = 12 THEN 'Meru'
        WHEN CAST(county AS INTEGER) = 13 THEN 'Tharaka-Nithi'
        WHEN CAST(county AS INTEGER) = 14 THEN 'Embu'
        WHEN CAST(county AS INTEGER) = 15 THEN 'Kitui'
        WHEN CAST(county AS INTEGER) = 16 THEN 'Machakos'
        WHEN CAST(county AS INTEGER) = 17 THEN 'Makueni'
        WHEN CAST(county AS INTEGER) = 18 THEN 'Nyandarua'
        WHEN CAST(county AS INTEGER) = 19 THEN 'Nyeri'
        WHEN CAST(county AS INTEGER) = 20 THEN 'Kirinyaga'
        WHEN CAST(county AS INTEGER) = 21 THEN 'Murang''a'
        WHEN CAST(county AS INTEGER) = 22 THEN 'Kiambu'
        WHEN CAST(county AS INTEGER) = 23 THEN 'Turkana'
        WHEN CAST(county AS INTEGER) = 24 THEN 'West Pokot'
        WHEN CAST(county AS INTEGER) = 25 THEN 'Samburu'
        WHEN CAST(county AS INTEGER) = 26 THEN 'Trans Nzoia'
        WHEN CAST(county AS INTEGER) = 27 THEN 'Uasin Gishu'
        WHEN CAST(county AS INTEGER) = 28 THEN 'Elgeyo-Marakwet'
        WHEN CAST(county AS INTEGER) = 29 THEN 'Nandi'
        WHEN CAST(county AS INTEGER) = 30 THEN 'Baringo'
        WHEN CAST(county AS INTEGER) = 31 THEN 'Laikipia'
        WHEN CAST(county AS INTEGER) = 32 THEN 'Nakuru'
        WHEN CAST(county AS INTEGER) = 33 THEN 'Narok'
        WHEN CAST(county AS INTEGER) = 34 THEN 'Kajiado'
        WHEN CAST(county AS INTEGER) = 35 THEN 'Kericho'
        WHEN CAST(county AS INTEGER) = 36 THEN 'Bomet'
        WHEN CAST(county AS INTEGER) = 37 THEN 'Kakamega'
        WHEN CAST(county AS INTEGER) = 38 THEN 'Vihiga'
        WHEN CAST(county AS INTEGER) = 39 THEN 'Bungoma'
        WHEN CAST(county AS INTEGER) = 40 THEN 'Busia'
        WHEN CAST(county AS INTEGER) = 41 THEN 'Siaya'
        WHEN CAST(county AS INTEGER) = 42 THEN 'Kisumu'
        WHEN CAST(county AS INTEGER) = 43 THEN 'Homa Bay'
        WHEN CAST(county AS INTEGER) = 44 THEN 'Migori'
        WHEN CAST(county AS INTEGER) = 45 THEN 'Kisii'
        WHEN CAST(county AS INTEGER) = 46 THEN 'Nyamira'
        WHEN CAST(county AS INTEGER) = 47 THEN 'Nairobi'
        ELSE 'Unknown'
    END AS county_name,
    "Sex",
    "Age",
    "Education",
    "Marital",
    "Latitude",
    "Longitude",
    "Quintiles",
    "indWeight",
    mobile_money_use,
    COALESCE(mobile_money_use, 0) AS mobile_money_use_clean,
    bank_use,
    COALESCE(bank_use, 0) AS bank_use_clean,
    mobile_money_access,
    "Access_Strand",
    "Savings_usage",
    "Loan_usage",
    "All_Insurance_including_NHIF",
    "Informal_usage",
    mobile,
    digital_acc,
    "Overall_Access_fnl",
    "Financial_literacy_index_fnl",
    finhealth_fnl,
    vul_index_fnl,
    CASE
        WHEN bank_use = 1 AND mobile_money_use = 1 THEN 'Fully Included (bank and mobile money)'
        WHEN mobile_money_use = 1 AND (bank_use IS NULL OR bank_use <> 1) THEN 'Partially Included (mobile only)'
        WHEN "Informal_usage" = 1
             AND (mobile_money_use IS NULL OR mobile_money_use <> 1)
             AND (bank_use IS NULL OR bank_use <> 1) THEN 'Informally Served (informal only)'
        ELSE 'Financially Excluded (none)'
    END AS financial_access_tier
FROM finaccess_master
WHERE county IS NOT NULL;
"""

with engine.begin() as conn:
    conn.execute(text(query))

print("View finaccess_clean created successfully.")


View finaccess_clean created successfully.


It preserves the original coded indicators, adds clearer county names, creates integer county codes for display, and adds helper variables that make later analysis easier to read.

The most important cleaning decision is transparency. The view does not hide the original `mobile_money_use` or `bank_use` values; it adds cleaned helper versions. It also embeds the financial access tier logic in one place so that later county, gender, and income comparisons use the same classification. This is the kind of reproducible data preparation that employers expect in professional analytics work.


### 8. Clean View Validation

The clean view is checked to confirm that the added county names, helper variables, and access tiers are available at respondent level. This validates the analysis layer used in later county summaries.


In [29]:
# Sample validation confirms the cleaned reporting layer is ready for later analysis.
query = """
SELECT
    interview__key,
    county,
    county_name,
    "Sex",
    "Age",
    mobile_money_use_clean,
    bank_use_clean,
    financial_access_tier
FROM finaccess_clean
ORDER BY interview__key
LIMIT 10;
"""

df_finaccess_clean_sample = pd.read_sql(query, engine)
df_finaccess_clean_sample

,interview__key,county,county_name,Sex,Age,mobile_money_use_clean,bank_use_clean,financial_access_tier
0,00-00-18-74,27,Uasin Gishu,1.0,4.0,1.0,1.0,Fully Included (bank and mobile money)
1,00-00-59-38,13,Tharaka-Nithi,2.0,5.0,2.0,3.0,Informally Served (informal only)
2,00-01-02-96,36,Bomet,1.0,2.0,1.0,1.0,Fully Included (bank and mobile money)
3,00-01-21-44,11,Isiolo,1.0,6.0,1.0,3.0,Partially Included (mobile only)
4,00-01-36-47,46,Nyamira,2.0,3.0,1.0,1.0,Fully Included (bank and mobile money)
5,00-01-41-64,34,Kajiado,1.0,3.0,1.0,1.0,Fully Included (bank and mobile money)
6,00-01-81-79,36,Bomet,2.0,3.0,1.0,1.0,Fully Included (bank and mobile money)
7,00-02-87-76,46,Nyamira,1.0,6.0,1.0,1.0,Fully Included (bank and mobile money)
8,00-03-02-06,8,Wajir,1.0,2.0,3.0,3.0,Financially Excluded (none)
9,00-03-71-00,29,Nandi,2.0,2.0,1.0,1.0,Fully Included (bank and mobile money)


The sample confirms that `finaccess_clean` works as intended. The output includes a clean integer `county` code, a readable `county_name`, the coded demographic fields, cleaned mobile money and bank helper columns, and the derived `financial_access_tier`. Validation against the database also confirms that the cleaned view contains `20,871` records, matching the master view because there were no null county values to exclude.

This completes the data filtering and cleaning section. The analysis has now documented missingness, confirmed the completeness of core segmentation variables, filtered working-age respondents, compared major urban centres, made null handling explicit, cleaned county display, and created a reusable analytical view. These steps strengthen the credibility of all later financial inclusion analysis because the data preparation decisions are visible, justified, and reproducible.


## Section Summary

The analysis confirms a complete national respondent base, broad county coverage, low missingness in core indicators, and a clean respondent-level view for reporting. The early results already show that mobile money is widespread, while formal access, income, and geography require deeper analysis.
